src/main/java/com/google/devtools/build/lib/remote/RemoteSpawnRunner.java

```diff
@@ -284,11 +300,20 @@ public SpawnResult exec(Spawn spawn, SpawnExecutionContext context)
                 Map<Digest, Message> additionalInputs = Maps.newHashMapWithExpectedSize(2);
                 additionalInputs.put(actionKey.getDigest(), action);
                 additionalInputs.put(commandHash, command);
+                Duration networkTimeStart = networkTime.getDuration();
+                Stopwatch uploadTime = Stopwatch.createStarted();
                 remoteCache.ensureInputsPresent(merkleTree, additionalInputs);
+                spawnMetrics.setUploadTime(
+                    uploadTime.elapsed().minus(
+                        networkTime.getDuration().minus(networkTimeStart)));
               }
               ExecuteResponse reply;
               try (SilentCloseable c = prof.profile(REMOTE_EXECUTION, "execute remotely")) {
+                // remove accounting for execution wait time
+                networkTime.disable();
```

Comment:
It seems cleaner to special case the execute method in the client interceptor.

Reply:
I'd be on board with that, let me see how it goes.

Reply:
Great. It should be quite straightforward. The gRPC stubs also expose a MethodDescriptor for each RPC.